File: tests/test_expense_management.py

import re
from playwright.sync_api import Page, expect

BASE_URL = "http://localhost:8000"

def wait_for_app_ready(page: Page):
    """
    Navigates to the base URL and waits for the app to signal it's ready.
    """
    page.goto(BASE_URL)
    page.locator('body[data-app-ready="true"]').wait_for(timeout=10000)

def test_add_and_delete_expense(page: Page):
    wait_for_app_ready(page)

    # 1. Create a group
    page.get_by_role("button", name="+").last.click()
    modal = page.locator("#new-group-modal.is-active").last
    modal.get_by_placeholder("e.g., Holiday Trip").fill("Expense Test")
    modal.get_by_placeholder("Comma-separated names").fill("Frank, Grace")
    modal.get_by_role("button", name="Save changes").click()

    # 2. Go to group details and add an expense
    page.get_by_role("link", name="Expense Test").click()
    page.get_by_role("button", name="Add Expense").click()
    expense_modal = page.locator("#add-expense-modal.is-active").last
    expense_modal.get_by_placeholder("e.g., Groceries").fill("Dinner")
    expense_modal.get_by_placeholder("e.g., 25.50").fill("50")
    expense_modal.locator('select[name="payer"]').select_option(label="Frank")
    expense_modal.get_by_role("button", name="Save Expense").click()

    # 3. Verify the balances
    balances = page.locator("#balances-summary")
    expect(balances.get_by_text("Frank is owed 25,00 €")).to_be_visible()
    expect(balances.get_by_text("Grace owes 25,00 €")).to_be_visible()

    # 4. Verify the expense is listed
    expense_list = page.locator("#expense-list")
    expense_box = expense_list.locator(".box", has_text="Dinner")
    expect(expense_box).to_be_visible()
    expect(expense_box.get_by_text("Amount: 50,00 €")).to_be_visible()
    expect(expense_box.get_by_text("Paid by: Frank")).to_be_visible()

    # 5. Delete the expense
    delete_button = expense_box.get_by_role("button", name="Delete")
    page.once("dialog", lambda dialog: dialog.accept())
    delete_button.click()

    # 6. Verify the expense is gone
    # After deletion, the entire list is re-rendered. We expect the placeholder.
    expect(page.get_by_text("No expenses recorded yet.")).to_be_visible()

    # 7. Verify balances are reset
    expect(balances.get_by_text("Frank is settled up")).to_be_visible()
    expect(balances.get_by_text("Grace is settled up")).to_be_visible()

File: tests/test_expense_management.js
const { test, expect } = require('@playwright/test');

const BASE_URL = "http://localhost:8000";

async function waitForAppReady(page) {
  await page.goto(BASE_URL);
  await page.locator('body[data-app-ready="true"]').waitFor({ timeout: 10000 });
}

test('add and delete expense', async ({ page }) => {
  await waitForAppReady(page);

  // 1. Create a group
  await page.getByRole('button', { name: '+' }).last().click();
  const modal = page.locator('#new-group-modal.is-active').last();
  await modal.getByPlaceholder('e.g., Holiday Trip').fill('Expense Test');
  await modal.getByPlaceholder('Comma-separated names').fill('Frank, Grace');
  await modal.getByRole('button', { name: 'Save changes' }).click();

  // 2. Go to group details and add an expense
  await page.getByRole('link', { name: 'Expense Test' }).click();
  await page.getByRole('button', { name: 'Add Expense' }).click();
  const expenseModal = page.locator('#add-expense-modal.is-active').last();
  await expenseModal.getByPlaceholder('e.g., Groceries').fill('Dinner');
  await expenseModal.getByPlaceholder('e.g., 25.50').fill('50');
  await expenseModal.locator('select[name="payer"]').selectOption({ label: 'Frank' });
  await expenseModal.getByRole('button', { name: 'Save Expense' }).click();

  // 3. Verify the balances
  const balances = page.locator('#balances-summary');
  await expect(balances.getByText('Frank is owed 25,00 €')).toBeVisible();
  await expect(balances.getByText('Grace owes 25,00 €')).toBeVisible();

  // 4. Verify the expense is listed
  const expenseList = page.locator('#expense-list');
  const expenseBox = expenseList.locator('.box', { hasText: 'Dinner' });
  await expect(expenseBox).toBeVisible();
  await expect(expenseBox.getByText('Amount: 50,00 €')).toBeVisible();
  await expect(expenseBox.getByText('Paid by: Frank')).toBeVisible();

  // 5. Delete the expense
  const deleteButton = expenseBox.getByRole('button', { name: 'Delete' });
  page.on('dialog', dialog => dialog.accept());
  await deleteButton.click();

  // 6. Verify the expense is gone
  await expect(page.getByText('No expenses recorded yet.')).toBeVisible();

  // 7. Verify balances are reset
  await expect(balances.getByText('Frank is settled up')).toBeVisible();
  await expect(balances.getByText('Grace is settled up')).toBeVisible();
});
